network/write_buffer.go

```diff
@@ -0,0 +1,167 @@
+package network
+
+import (
+	"context"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/golang/snappy"
+	"github.com/grafana/walqueue/types"
+	"go.uber.org/atomic"
+)
+
+// writeBuffer handles buffering the data, keeping track if there is a write request already running and kicking off the
+// write request as needed. All methods need to be called in a thread safe manner.
+type writeBuffer[T types.Datum] struct {
+	items []T
+	// writeAvailable keeps track if there is a write request going out.
+	writeAvailable *atomic.Bool
+	wrBuf          []byte
+	snappyBuf      []byte
+	log            log.Logger
+	cfg            types.ConnectionConfig
+	stats          func(stats types.NetworkStats)
+	isMeta         bool
+}
+
+func newWriteBuffer[T types.Datum](cfg types.ConnectionConfig, stats func(networkStats types.NetworkStats), isMeta bool, l log.Logger) *writeBuffer[T] {
+	return &writeBuffer[T]{
+		items:          make([]T, 0),
+		writeAvailable: atomic.NewBool(true),
+		cfg:            cfg,
+		stats:          stats,
+		isMeta:         isMeta,
+		log:            l,
+	}
+}
+
+// ForceAdd is only used when we need to force items to the queue, this is generally done as part of a config change.
+func (w *writeBuffer[T]) ForceAdd(item T) {
+	w.items = append(w.items, item)
```

Comment:
Should this check for BatchCount -> Send like Add does?

Reply:
Sounds good